### PR TITLE
Implement missing virtual tables for Accord debugging

### PR DESCRIPTION
**Pull request by iamaleksey**
patch by Aleksey Yeschenko; reviewed by Benedict Elliott Smith for CASSANDRA-20062

### CODE REVIEW

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: so, there was a thread about vtable keyspaces being outside the default one (the context was the metric one) and people are strongly against adding new virtual keyspaces and want all vtables in the existing keyspace.

File: src/java/org/apache/cassandra/schema/SchemaConstants.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/schema/SchemaConstants.java]: since the code can walk these the fact `VIRTUAL_ACCORD_DEBUG ` is conditional makes using this brittle... its been a source of pain for tables with places like guardrails as their counts are wrong.  And when I did the last rebase it was something I had to worry about as the accord stuff was present but might not exist...

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: This one is purely for *debugging* Accord (mostly by Accord developers), not for consumption by others, and is 'unstable'. Definitions can change at any point, and I don't want to put unstable/debug tables into a 'stable' system_views keyspace.

File: src/java/org/apache/cassandra/schema/SchemaConstants.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/schema/SchemaConstants.java]: Yeah. This was also missing `VIRTUAL_VIEWS` though, which is not conditional. And `LOCAL_SYSTEM_KEYSPACE_NAMES` above *does* include `system_accord` unconditionally despite the keyspace itself only added conditionally. Need to make it consistent in *some* direction. I'll think about it.

Comment (iamaleksey) [reply src/java/org/apache/cassandra/schema/SchemaConstants.java]: I think the sets should include these keyspace names, but we can't make it conditional without referring to `DatabaseDescriptor` in `SchemaConstants`. And that would defeat the purpose of `SchemaConstants` (supposed to be usable without instantiating DD).

From another angle, why not instantiate `system_accord` and `system_accord_debug` unconditionally? We do expect Accord to be used heavily, and even for those not using it, there is little harm in those keyspaces being present but not used. No real harm done.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: why this view point?  the vtables that are in cep-15-accord have been for developers and operators to better understand what is going on... so "not for consumption by others" is new to me.

> and is 'unstable'

I brought this up in dev@ awhile back... we don't actually have the tools to define that right now in C* so anything that is published can't be changed... we need to address exposing tables as breakable first.

File: src/java/org/apache/cassandra/schema/SchemaConstants.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/schema/SchemaConstants.java]: so, one line of thinking is how this all works with downgrades.  We *require* 5.0 to be able to downgrade to 4.1... the main area that this breaks is when we make schema changes...  for the accord keyspace we don't do anything fancy but the pattern of delaying till you opt in does help for downgrade.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: We *cannot* provide these vtables with a stable API - they expose internal implementation details, intentionally, details we expect to change (even in a minor release). They should be members of a keyspace we can clearly document as unstable, so that users do not take a programmatic dependency upon them.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: why this change?  This is far harder to use than the string version, as C*/Accord always reports the string version.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: same as above, why are we moving away from the string version that gets reported to users/operators?

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: same as above, the tuple is far harder to work with than the string version.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:537]: > Fix up existing Accord debug virtual tables

is this the fixup?  Moving off UDT to Tuple?  I am not opposed to this as I think java driver was iffy about handling the UDT.  I am wondering what is the main motivation behind this?

Again, not against, just trying to understand.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:568]: this is actually a bug... I was going to fix in CASSANDRA-20054 but was asked to drop to speed up the review.  We have a concept of "historic transactions", so those come back as `null` here.

The fix for it is

```
-                Invariants.checkState(reason == Reason.Self, "Txn %s unknown for reason %s", txnId, reason);
+                Invariants.checkState(reason != Reason.Self, "Self Txn %s is unknown", txnId);
+                // Self txn depends on this txn, but it is not known...
+                ds.row(userTxn.toString(), shard.storeId, depth, txn.txnId.toString(), reason.name());
+                ds.column("save_status", "<unknown>");
```

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: why is `CoordinationStatus` dropped?  I use this for debugging deployed clusters...

**Review by dcapwell — CHANGES_REQUESTED** [review]
`txn_blocked_by` is used by tests for debugging and the logic was not updated to the breaking changes that are in this patch.  What is dangerous is that the tests passing doesn't tell you this, we only do this when things are broken, so we only fail when we need things the most.

I am not sure why we made 2 critical changes

1) switch to isolated keyspace.  This does get pushback in trunk today where people are strong saying no to this.  The argument given so far is that we need to be able to break things, but that doesn't exist in C* today so any table published is part of the public API, if we want breakable tables we need to work on that in C* first.

2) `txn_blocked_by` moving to tuples is a UX issue.  C*/Accord always reports the string version and switching to tuple makes it so much harder for people to debug.  We need to take the string, and write java code to figure out what the tuple should look like... this has been a huge pain point of `accord_coordination_status`.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: > They should be members of a keyspace we can clearly document as unstable, so that users do not take a programmatic dependency upon them.

this is not enough to not be part of the public API.  I brought this up when finding vtables were getting added that are super brittle or with bad UX and how we can fix them without doing breaking changes.  We do not have experimental tables at the moment and we need to flesh that out as a project first if we want that.  Putting the word `debug` in the keyspace is not enough for that, and if that is a goal I am against this patch; we need to fix this in the project first.

People in dev@ are not opposed to the idea, but we need to flesh that out and document things.

Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: There’s a difference between poorly design UX and UX specifically documented as exposing internal details that cannot be relied upon (even between patch versions). To be clear, these tables aren’t *ever* expected to be used for programmatic access. This isn’t the same as “experimental” - this is “WE MAKE NO PROMISES EVER”.

I would also be fine with having this whole keyspace hidden behind some config property for now, so you have to enable it and read some documentation making this clear. This seems to sidestep any issue and we can revisit exposing it by default once broader discussion has happened.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: How was it of value? The initial coordination should last at most 10s (until it times out), after which all the relevant state for debugging in progress transactions should be found in the progress log table.

We could perhaps enrich the progress log home state to include information about active coordinations, including more useful information such as the phase we have reached. This would provide a more consistent means of monitoring this.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [inline src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: In theory a `TxnId` can have other flags. This is not used at the moment, but we might want to future-proof this, and report the raw flags as well? An example might be the "coordinator optimisation" that permits smaller quorums at the expense of slightly more coordinator work, that we expect to implement before the new year (and for which I think the easiest way to configure it would be with a `TxnId` flag) - but no doubt there will be other flags we want to introduce later.

**Review by belliottsmith — APPROVED** [review]
one minor suggestion, that would be preferable to address before merge. +1

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: Alternatively, reporting the `String` representation as suggested by David isn't a bad idea really, and removing all of the UDT/Tuple across the board for Timestamp/TxnId types.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:537]: UDT support for virtual keyspaces was added incorrectly, and adding it correctly is a bit out of scope for this ticket. So I switched from UDT to a tuple - the UDT wasn't really providing much utility either, it's just two elements that are pretty easily human-readable in a tuple without field names.

Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:537]: Adding virtual UDTs correctly would involve:
1. Adding a virtual table, `system_virtual_schema.types`, to expose all virtually-registered UDTs per virtual keyspace
2. Modifying the drivers to read virtual UDT definitions from this table

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: That ^.

I guess I'm fine with hiding it behind a property. Will update.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: Fair enough. Got rid of all timestamp/txnId tuples.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: Addressed.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: All addressed.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:568]: Just `txnId.toString()` I assume? Making this change breaks all the tests in `AccordDebugKeyspaceTest` though.

File: src/java/org/apache/cassandra/schema/SchemaConstants.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/schema/SchemaConstants.java]: If someone downgrades, the worst thing that will happen is they'll have some remnant Accord keyspace files with no data?

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java]: Addressed by switching all timestamps and txnIds to `text`.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: > this is “WE MAKE NO PROMISES EVER”.

This isn't a thing in the project today, so anything exposed is now part of the public API.  This is why I am harping on this topic as its been brought up... If you want tables/keyspaces where you can break w/e you wish, we need to *add* such a thing into the project, which I strongly do not feel this patch does.

> I guess I'm fine with hiding it behind a property

If we wish to hid the new progress log tables thats fine, but I am a hard -1 on hiding the existing tables as they were designed for operators / developers to help manage a production cluster.  They are not "WE MAKE NO PROMISES EVER".

File: src/java/org/apache/cassandra/schema/SchemaConstants.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/schema/SchemaConstants.java]: hope so... I pointed out that 5.0 was broken and the person who was supposed to take ownership of downgrade was supposed to work on it... I don't know if the test I used to show downgrade was broken got merged or not... so no clue if we have any coverage =(

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:537]: as the person who added UDT I am fine switching to tuple, I just wanted to confirm this was the issue

Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:537]: > the UDT wasn't really providing much utility either

It was nice in python-dtest but was annoying in java driver... but yeah I am cool with the tuple migration away from UDT

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:568]: its fine to leave as is and ill fix this later

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: > How was it of value?

It had the same value as the other nodetool commands we use to see what queries are actively running.  When I see user latencies start to regress or fall off a wall I have been using it to see what is going on and what is in-flight.

> We could perhaps enrich the progress log home state to include information about active coordinations, including more useful information such as the phase we have reached. This would provide a more consistent means of monitoring this.

I think we need a table that tells operators what is going on, and isn't expected to break in any random release... I agree this table has the 10s weakness you point out, but we need something.  If the progress log tables are expected to break w/e then we don't have the visibility we need.

File: src/java/org/apache/cassandra/config/AccordSpec.java
Comment (dcapwell) [inline src/java/org/apache/cassandra/config/AccordSpec.java]: this documentation is not enough for saying the tables are not exposed to the public API.  

See `Re: [EXTERNAL] Re: [DISCUSS] Adding experimental vtables and rules around them` for the dev thread about this topic.

To unblock you I am fine withs something like `enable_unsupported_virtual_tables` as the name tells you they are `unsupported`, but I am a hard -1 on merging Accord to trunk with this topic not dealt with.  As a project we need to define a way to have experimental (aka unsupported) tables, and until that is done Accord can not merge to trunk as it is the first to define such tables...

Comment (iamaleksey) [reply src/java/org/apache/cassandra/config/AccordSpec.java]: These are neither experimental nor unsupported. They are purely internal. I don't know how much clearer one can be.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: If we’re throwing hard -1s around, I’m a hard -1 on releasing any of these as production tables, including the ones that had merged to cep15 already. They leak implementation details and are unsuitable to be treated as public APIs

File: src/java/org/apache/cassandra/config/AccordSpec.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/config/AccordSpec.java]: To to be as explicit as possible: these tables are intended for human consumption only. To my knowledge it is not project policy that virtual tables must be consumable by applications, but even if you take that they are then flagging them off resolves any concern. This is all we mean by API-unstable. Really we mean "NOT AN API, ONLY AN HMI". Applications that take a dependency on them can expect to break without warning, but human operators can generally expect to interpret them if they understand the relevant internals.

We will never provide an API for viewing _implementation details_ while I have a say, as it would prevent reasonable development. But, it is helpful to expose these details to operators for human interpretation of the cluster state _only_.

I am aware of no project policy that would be in conflict with merge. So, if you are opposed to merging CEP-15 once this has merged, please decide what you require to proceed and take it to the list for discussion.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: > They leak implementation details

I can't speak to migration as I don't know it well, but the 2 used tables are

1) coordination
2) txn block

for coordination it mimics the existing nodetool commands to see what CQL is running, so it shows what the txn is and how long its been coordinated.  What implementation detail are you concerned about here?

For the txn blocking one, this is a thing operators need as it shows what blocks what and why.  The API intentionally only leaks txn vs key dependencies and not much else; so what are your concerns about the implementation details?  I am also curious why you never raised these concerns when you reviewed the blocking patch; your feedback there made the table far better.

Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: The coordination table as discussed already I consider to be pretty useless, much less so than the one for long running CQL commands as we explicitly time these out and remove them within 10s, there is nothing that can be shown to help an operator. I did not review this table. I do absolutely support enriching the progress log table with this information (in fact, a more useful version of this, to include the phase of coordination and perhaps even the number of responses we have for that phase)

The txn block table leaked actual cache references until my recent patch, so clearly I didn't review it very carefully :/ 
But this leaks at least: SaveStatus, ExecuteAt, StoreId, but also leaks details about how we manage execution dependencies (which has already been drastically modified more than once during development).

I did not anyway honestly consider this was intended to be consumed by applications, nor was I aware of any controversy about human-only virtual tables.

These are all hugely valuable tables, they just should not be considered stably machine readable.

File: src/java/org/apache/cassandra/config/AccordSpec.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/config/AccordSpec.java]: >  But, it is helpful to expose these details to operators for human interpretation of the cluster state only.

If the assumption is to expose this for operators, then operators will build tools as manual work is a massive waste of operator time.  So breaking the table becomes a massive burden for operators, who you are saying are a target audience.  

If the audience is tests, then flagging on for tests and off for prod is fine for me, we do this with system properties today.

> To my knowledge it is not project policy that virtual tables must be consumable by applications,

at no point have I said "consumable by applications", I said "public api".  We have had a long stance that every table and every JMX is part of our public API and we can not break.  I have also said there is a desire and willingness to define breakable tables, but we must do it in a way that actually documents that tables are breakable, which is one of the main reasons the ML has focused on how do we expose this detail to users / operators.  If someone sees `debug` do they think this can change at a moments notice, or that this is useful to debug the issues you are facing?  The focus on the ML has been the word "experimental" as the focus was on tables not 100% ready for production, but this patch wants a different definition, one I don't think anyone is opposed to... my concern is articulating this to people, I want to avoid people being blind sided by a breaking change.

Tl;DR - if this is for tests or people working on accord patches and won't be enabled for prod, I am mostly fine, but still think a better config name would help give this message.

If the audience is operations this then means enabled on prod, then I don't think the config name or documentation tells operators what rules they should expect around the API changing.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (iamaleksey) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: > I think we need a table that tells operators what is going on, and isn't expected to break in any random release... I agree this table has the 10s weakness you point out, but we need something. If the progress log tables are expected to break w/e then we don't have the visibility we need.

I think we can and should expose something like that before final release, but maybe delay until we are closer to completion and things are more-or-less fully settled? I've kept a skeleton `AccordVirtualTables` class to put in whatever tables we want to expose to the wider audience as final and stable later.

File: src/java/org/apache/cassandra/config/AccordSpec.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/config/AccordSpec.java]: I was previously an operator, and manually using introspection views on e.g. SQL Server was a perfectly normal activity, and I did not use scripts to access this information. By "Consumable by applications" I meant "API" (the A means application)

I'm obviously not _opposed_ to folk using scripts, just so long as they accept the possibility that their scripts may need tweaking when they go to use them, because we don't promise not to break them. If the scripts are just labour saving devices that the operator understands, this shouldn't be overly onerous, or destroy the value proposition.

The alternative in my view is either _not providing these views_, or providing ones with massively reduced information. One option might be to have restricted versions of each view, and a config flag that enriches the tables with more information, and makes more tables available. But I would prefer to explore that closer to release of 5.1, as in the meantime many of these tables will change anyway.

I'm agnostic regarding naming of any flag, and also perfectly supportive of migrating this to some endorsed "unstable API" mechanism as and when it becomes available.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: >  but maybe delay until we are closer to completion and things are more-or-less fully settled?

I am fine with that.  If we want to say we want a coordination table but we are not yet ready to implement one, then I am fine removing in this patch.

File: src/java/org/apache/cassandra/config/AccordSpec.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/config/AccordSpec.java]: > The alternative in my view is either not providing these views, or providing ones with massively reduced information

I am fine with the progress log tables being off by default and treating it as testing code.  I am 100% fine saying we will flesh out an operator focused table closer to merge or release, and that isn't required scope of this patch.

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: I’m fine quickly enriching the progress log table before merge. One less thing to worry about

Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: I’ll try to add that tonight

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: > The coordination table as discussed already I consider to be pretty useless,

stated in another comment, I am fine saying we will punt on this table closer to the release and drop it for now.  What is running is very useful but if you both don't feel we are ready for that we can drop for now.

> The txn block table... But this leaks at least: SaveStatus, ExecuteAt, StoreId, but also leaks details about how we manage execution dependencies (which has already been drastically modified more than once during development).

this table is critical for someone to answer the question "why did my txn timeout", as it shows what is blocking it still.  If we say key vs txn dependencies should not be exposed then thats mostly fine as the txn_id matters the most... if you are blocked on X you then can ask why is X blocked and keep going until you find what is stuck. But if you don't expose save status and timestamps then what can you meaningfully say why something is stuck?

So, if you prefer to drop the key vs txn part, I am fine removing in this patch. If you feel the txn id and save status is too much, then we have problems as I don't know how to answer the above question.

> I did not anyway honestly consider this was intended to be consumed by applications, nor was I aware of any controversy about human-only virtual tables.

when ever there has been a patch to alter any existing table the same conversation comes up about backwards compatibility; once it reaches the public API we have to maintain it.  We have 0 tables where this stance doesn't hold true today, this patch is arguing to be the first.

As stated in the other thread, I am fine with a view point that these tables are for testing so off by default and off in prod, but exist to aid developers writing patches or to improve test errors; there is clear value in that.

Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:81]: Again, this is not about whether this should be enabled in prod, and I don’t really understand why that is a consideration. Anybody should feel comfortable enabling this in prod if they understand the guarantees.

Also again, an API means an application programming interface. I agree public APIs need to be maintained in a given way, I disagree that all public interfaces are APIs. 

We’re going round in circles here, but it sounds like we can shelve further discussion on this until later.

**Comment by dcapwell** [comment]
The main thing blocking my +1 is the `txn_blocked_by` table.  We either need to move it back where it was, or change JVM-Dtest to handle the changes.

My stance is `txn_blocked_by` should move back and if there are concerns about the table we should address them.  If there is a valid believe that this table should not be exposed in prod in the current state I am fine moving it (though I disagree).

File: src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java:91]: It is a bit more complicated than I had expected to do a decent job of this, so I'm going to add a ticket to track adding this in future.

**Review by dcapwell — APPROVED** [review]

**Pull request merged**
